drivers: led: max7219: add driver and sample

yanagiis opened the pull request:
This PR implements MAX7219 driver and provides sample.

yanagiis committed:
drivers: led: max7219: implement driver

This commit implements the LED interface for the Maxim MAX7219 LED driver.

Signed-off-by: Jimmy Ou <yanagiis@gmail.com>

yanagiis committed:
drivers: led: max7219: add sample

This commit provides a MAX7219 sample for demonstration.

Signed-off-by: Jimmy Ou <yanagiis@gmail.com>

yanagiis committed:
drivers: led: max7219: add build_all led test

Add build_all led test.

Signed-off-by: Jimmy Ou <yanagiis@gmail.com>

galak commented on drivers/led/Kconfig.max7219:
```diff
@@ -0,0 +1,20 @@
+
+menuconfig MAX7219
+	bool "MAX7219 LED drivers"
+	depends on SPI
```
Add a `default $(dt_compat_enabled,$(DT_COMPAT_MAXIM_MAX7219))`

(see `drivers/sensor/lm75/Kconfig` for an example)

yanagiis replied:
OK. Thanks. I will add that.

galak commented on tests/drivers/build_all/led/prj.conf:
```diff
@@ -2,3 +2,4 @@ CONFIG_TEST=y
 CONFIG_TEST_USERSPACE=y
 CONFIG_LED=y
 CONFIG_LED_GPIO=y
+CONFIG_MAX7219=y
```
This should be able to be removed when the `default` in Kconfig is added.

galak commented on samples/drivers/led_max7219/prj.conf:
```diff
@@ -0,0 +1,4 @@
+CONFIG_LED=y
+CONFIG_LOG=y
+CONFIG_MAX7219=y
```
can be removed when `default` added to Kconfig.

gmarull commented on drivers/led/Kconfig.max7219:
```diff
@@ -0,0 +1,20 @@
+	  Enable driver for the Maxim MAX7219 SPI LED controllers.
+	  It is supporting up to 64 individual LEDs.
+
+config MAX7219_INITIALIZATION_DELAY
+	int "MAX7219 initialization delay in milliseconds"
+	depends on MAX7219
+	default 0
+	range 0 2147483647
+	help
+	  This option delays MAX7219 initialization for the specified
+	  amount of milliseconds. This is used to allow MAX7219 get ready
+	  before initialing.
```
why is this needed?

yanagiis replied:
MAX7219 may not be ready when zephyr calls `max7219_led_init`.
Plus it only has MOSI pin, so I can't check its status.
I have no idea how to handle this situation in zephyr. 
Do you have any suggestions?

galak reviewed (changes requested):
A few changes here and there.

gmarull commented on drivers/led/max7219.c:
```diff
@@ -0,0 +1,252 @@
+	struct max7219_data *data = dev->data;
+	int ret;
+
+#if defined(CONFIG_MAX7219_INITIALIZATION_DELAY) && CONFIG_MAX7219_INITIALIZATION_DELAY > 0
+	k_msleep(CONFIG_MAX7219_INITIALIZATION_DELAY);
+#endif
```
why? this will delay the whole system init

gmarull commented on drivers/led/max7219.c:
```diff
@@ -0,0 +1,252 @@
+ */
+
+#define MAX7219_INDEX(led) ((led) & (~BIT_MASK(6))) >> 6
+#define DIGIT_INDEX(led)   (((led)&BIT_MASK(6)) >> 3)
+#define SEGMENT_INDEX(led) ((led)&BIT_MASK(3))
```
` & `

gmarull commented on dts/bindings/led/maxim,max7219.yaml:
```diff
@@ -0,0 +1,55 @@
+    num-cascading:
+      type: int
+      required: false
+      default: 1
```
ditto

yanagiis replied:
OK. MAX7219 has a default value. I should remove the default value. Thanks a lot.

gmarull commented on samples/drivers/led_max7219/sample.yaml:
```diff
@@ -0,0 +1,7 @@
+  name: MAX7219 sample
+tests:
+  sample.drivers.led.max7219:
+    platform_allow: stm32_min_dev_black
```
please filter based on `dt_compat_enabled`

yanagiis replied:
I will add it. Thanks.

gmarull commented on samples/drivers/led_max7219/src/main.c:
```diff
@@ -0,0 +1,63 @@
+#include <zephyr/sys/util.h>
+#include <zephyr/zephyr.h>
+
+#define LOG_LEVEL CONFIG_LOG_DEFAULT_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(app);
```
please define app level log level. don't hardcode it.

yanagiis replied:
OK. I will change it. Thanks.

gmarull commented on samples/drivers/led_max7219/src/main.c:
```diff
@@ -0,0 +1,63 @@
+
+void main(void)
+{
+	const struct device *led_dev = DEVICE_DT_GET_ANY(maxim_max7219);
+	int i, ret;
+
+	if (led_dev == NULL) {
+		LOG_ERR("No device with compatible maxim,max7219 found");
+		return;
+	}
```
why not `DEVICE_DT_GET_ONE`?

yanagiis replied:
It's better than `DEVICE_DT_GET_ANY`. Thanks for your suggestion. I will change it.

yanagiis commented on dts/bindings/led/maxim,max7219.yaml:
```diff
@@ -0,0 +1,55 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  Maxim MAX7219 display driver
```
OK. Thanks. My mistake. I will remove it.

yanagiis committed:
drivers: led: max7219: remove blank line

Should follow zephyr coding-sytle.

Signed-off-by: Jimmy Ou <yanagiis@gmail.com>

yanagiis committed:
drivers: led: max7219: add dt_compat_enabled

Add dt_compat_enabled for auto enable MAX7219.

Signed-off-by: Jimmy Ou <yanagiis@gmail.com>

yanagiis committed:
drivers: led: max7219: remove multi-line symbol

The description just one line sentence. Remove the multi-line symbol.

Signed-off-by: Jimmy Ou <yanagiis@gmail.com>

yanagiis committed:
drivers: led: max7219: use spi_is_ready instead of device_is_ready

Signed-off-by: Jimmy Ou <yanagiis@gmail.com>

yanagiis commented on drivers/led/max7219.c:
```diff
@@ -0,0 +1,252 @@
+	k_msleep(CONFIG_MAX7219_INITIALIZATION_DELAY);
+#endif
+
+	if (!device_is_ready(config->spi.bus)) {
```
OK. Thanks. I will change it.

yanagiis commented on drivers/led/max7219.c:
```diff
@@ -0,0 +1,252 @@
+	k_msleep(CONFIG_MAX7219_INITIALIZATION_DELAY);
+#endif
+
+	if (!device_is_ready(config->spi.bus)) {
```
Thanks. I will change it.

yanagiis commented on drivers/led/max7219.c:
```diff
@@ -0,0 +1,252 @@
+#include <zephyr/sys/util.h>
+#include <zephyr/zephyr.h>
+
+#define LOG_LEVEL CONFIG_LED_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(max7219);
```
OK, thanks. I will add it.

yanagiis commented on dts/bindings/led/maxim,max7219.yaml:
```diff
@@ -0,0 +1,55 @@
+    scan-limit:
+      type: int
+      required: false
+      default: 7
```
OK. I think I should remove the default value. Thanks a lot.

yanagiis commented on dts/bindings/led/maxim,max7219.yaml:
```diff
@@ -0,0 +1,55 @@
+    intensity:
+      type: int
+      required: false
+      default: 0
```
OK. MAX7219 has a default value. I should remove the default value. Thanks a lot.

yanagiis committed:
drivers: led: max7219: add dt_compat_enabled

Add filter dt_compat_enabled with maxim,max7219.

Signed-off-by: Jimmy Ou <yanagiis@gmail.com>

yanagiis committed:
drivers: led: max7219: remove LOG_LEVEL define

Remove LOG_LEVEL define. Use LOG_MODULE_REGISTER with level directly.

Signed-off-by: Jimmy Ou <yanagiis@gmail.com>

yanagiis committed:
drivers: led: max7219: remove LOG_LEVEL in sample

Signed-off-by: Jimmy Ou <yanagiis@gmail.com>

yanagiis committed:
drivers: led: max7219: use DEVICE_DT_GET_ONE

DEVICE_DT_GET_ONE check the device at compile time.

Signed-off-by: Jimmy Ou <yanagiis@gmail.com>

yanagiis commented:
Hi @simonguinot. Sorry, I don't get it.
This module is what I use.
[MAX7219 8-Digital Segment Digital LED Display](https://www.amazon.com/HiLetgo-MAX7219-8-Digital-Segment-Digital/dp/B01D0WSCJA/ref=sr_1_5?keywords=max7219&qid=1657126308&sr=8-5)
I thought it connects to LEDs?

yanagiis committed:
drivers: led: max7219: remove default value

User should set properties in the dts.

Signed-off-by: Jimmy Ou <yanagiis@gmail.com>

yanagiis committed:
drivers: led: max7219: formatt codes

Follow zephyr clang-format style.

Signed-off-by: Jimmy Ou <yanagiis@gmail.com>

yanagiis committed:
drivers: led: fix build_all link error

Test build_all report link error which related to GPIO and SPI devices.
Add depends on GPIO and SPI in testcase.yaml.

Signed-off-by: Jimmy Ou <yanagiis@gmail.com>

simonguinot commented:
> Hi @simonguinot. Thanks for your review.
> 
> Sorry, I don't get it. This module is similar to what I use. [MAX7219 8-Digital Segment Digital LED Display](https://www.amazon.com/HiLetgo-MAX7219-8-Digital-Segment-Digital/dp/B01D0WSCJA/ref=sr_1_5?keywords=max7219&qid=1657126308&sr=8-5) I thought it connects to LEDs?

Hi @yanagiis. I think I misled you by speaking of LCD. It is my mistake.

Let me clarify: the MAX72{19,21} is a display LED driver. It displays digits. And the LED API is not suited to display digit values.

But maybe you can use a display driver to display the digits ? See the drivers/display directory. It looks like there are LED matrix drivers among them.

I don't know at all the display API. So maybe I am all wrong and the MAX72{19,21} is not compatible. But please have to look.

nashif commented on samples/drivers/led_max7219/src/main.c:
```diff
@@ -0,0 +1,63 @@
+/*
```
do we really need yet another copy of this sample for another driver? This seems to the same as led_lp3943. Either unify and make existing sample work for new driver or provide something that is new.

yanagiis commented:
Hi @simonguinot. 
I have taken look at the display driver and `nrf_led_matrix` driver and related [issue](https://github.com/zephyrproject-rtos/zephyr/issues/40208).

IMHO, the `Display API` is also not suited to display digit values when MAX7219 connects to 7-segment digits LED display.
But the `Display API` is more suitable than the led API when MAX7219 connects to the LED matrix.

I will rewrite it as a display driver.
Thanks again for your suggestion.

simonguinot commented:
> Hi @simonguinot. I have taken look at the display driver and `nrf_led_matrix` driver and related [issue](https://github.com/zephyrproject-rtos/zephyr/issues/40208).
> 
> IMHO, the `Display API` is also not suited to display digit values when MAX7219 connects to 7-segment digits LED display. But the `Display API` is more suitable than the led API when MAX7219 connects to the LED matrix.
> 
> I will rewrite it as a display driver. Thanks again for your suggestion.

No problem !
